### PR TITLE
WIP: Container annotation manager.

### PR DESCRIPTION
**Pull request by Integral**
This is not yet ready to merge, but in the same state as it was in February with all new changes.
Please review.

### CODE REVIEW

File: packages/container-annotation/ContainerAnnotationManager.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+          // conntainer annotation got updated
+          if(this._state.annotations[nodeId]) shouldUpdate = true
+        }
+      }
+    })
+
+    // Compute range for created/removed annos
+    if(dirtyNodes.length === 2) {
+      // TODO: We should get container via API
+      const container = this.doc.get('body')
```
Comment (obuchtala) [inline packages/container-annotation/ContainerAnnotationManager.js:104]: We need to fix this.
I need to take a closer look at this method. Maybe I have an idea.

Comment (obuchtala) [reply packages/container-annotation/ContainerAnnotationManager.js:104]: Will let u know if I find something.

Comment (Integral) [reply packages/container-annotation/ContainerAnnotationManager.js:104]: Also maybe we should move node range computation to container annotation model.

**Comment by michael** [comment]
@Integral please in future send separate PR for things like SnapshotEngine fix, then I'm able to get them into develop faster.

**Comment by Integral** [comment]
@michael i don't know what it's doing here... i should remove this, because it was fixed by me right in develop branch.

**Comment by michael** [comment]
@oliver---- we are publishing this as an npm release `1.0.0-container-annos.1` and `npm publish --tag container-annos`

However it would be good to get this into develop. Do you think it would affect stability.. i mean if no container annos are present in the document.

**Comment by obuchtala** [comment]
I guess we want this merged sooner than later, to be able to find out :)

**Comment by michael** [comment]
So should we do it? :)

**Comment by obuchtala** [comment]
I guess so :)

**Comment by michael** [comment]
Anything to clean up or review before we do it?

**Comment by obuchtala** [comment]
There are still some brutal hackz, like `doc.get('body')` hard-coded.
We need to fix that first :(

**Comment by Integral** [comment]
@oliver---- you was going to think about it ;)

**Comment by michael** [comment]
Hmm... how to best get rid of that hardcoded body?

**Comment by obuchtala** [comment]
Need to understand the implementation. But this is not what will happen instantly.

**Comment by obuchtala** [comment]
Let us put it on the agenda.

**Comment by obuchtala** [comment]
We need to rewrite `ContainerAnnotationManager._onDocumentChange()`.
I have started to do some tidy-up in that class, to increase comprehensibility.
Now we have to identify a viable strategy how to detect changes that affect the indexes maintained by
this class.
And we need a test-case for this.

**Comment by obuchtala** [comment]
There are tests in `ContainerAnnotation.test.js`

**Comment by obuchtala** [comment]
I left some comments:
>   This class maintains a set of Markers (see MarkersManager)
> where each marker represents a fragment of a ContainerAnnotation.
>
> It needs to react on DocumentChanges whenever:
> - a ContainerAnnotation has been created, deleted,
>   or updated
> - a node has been inserted or removed from a container
>   being covered by ContainerAnnotation

>  TODO:
>  - try to complement the requirement description started in the class header
>  - try find a reasonable implementation of the description
>  - maintain an explanation of the taken approach here, so that
>    it is easier to understand and verify the implementation in future

**Comment by Integral** [comment]
@oliver---- we don't want to merge until those todos will be done?

**Comment by obuchtala** [comment]
Nope. Sorry.

**Comment by obuchtala** [comment]
Because the applied hackz are so brutal, that we must solve the proposed TODOs anyway

**Comment by Integral** [comment]
Fine for me as we did npm release anyways and i will not be able to dive into details until Friday.

**Comment by obuchtala** [comment]
Ok

**Comment by obuchtala** [comment]
Some ideas:
- Detecting changes to ContainerAnnotations should be rather easy: there are several ways to approach this, either via inspecting `change.ops`, or by checking if `change.updated` or `change.deleted` affect known annos, plus looking into `change.created` if there is container anno. Probably the first is easier.
- Detecting changes to Containers is probably more involved: on the highest level (without lists) it is rather easy by detecting updates to containers. For lists, we would need to identify updates to lists and navigate up to the container.

**Comment by obuchtala** [comment]
Obsolete.

**Pull request closed without merging**